Feature: Ability to change flow ratios by extrusion path types (Code + Docs)

valerii-bokhan opened the pull request:
# Description

This new feature allows users to override flow ratios for the following extrusion path types:
- First layer (excluding Brims and Skirts)
- Outer and inner walls
- Overhang perimeters
- Sparse infill
- Internal solid infill
- Gap fill
- Support and Support interfaces

If the checkbox is disabled in UI, all the new fields will be hidden, and the values will not be applied.

**Use case example:** My goal was to eliminate the "wrinkles effect" on the outer walls when printing with low layer heights, such as, e.g., 0.04mm. Instead of increasing the temperature, I adjusted the flow ratio for the outer walls only to address the issue.

# Screenshots

How UI looks when the feature is disabled:
<img width="392" height="264" alt="image" src="https://github.com/user-attachments/assets/95a53671-1c97-4a58-86b4-bb12650077d8" />

How UI looks when the feature is enabled:
<img width="397" height="530" alt="image" src="https://github.com/user-attachments/assets/dca977dc-84b1-4536-a054-8a2c2c5295b4" />

## Tests

I tested it by checking the values in the slicer (that the multiplicators were applied). Additionally, I tested it on my Bambu Lab A1 mini, printing before and after values with overriding.

valerii-bokhan committed:
Feature: Ability to change flow ratios by extrusion path types (Code + Docs)

This new feature allows users to override flow ratios for the following extrusion path types:

* First layer (excluding Brims and Skirts)
* Outer walls
* Inner walls
* Overhang perimeters
* Sparse infill
* Internal solid infill
* Gap fill
* Support
* Support interfaces

valerii-bokhan commented:
> Thank you, I’ve been waiting for this addition for a long time and I’m testing it. When you say overwrite, you mean it doesn’t take into account the overall flow multiplier, right? Personally, I believe it’s better to follow the same logic as the flow multiplier for the lower and upper surfaces, so that it multiplies the two flow values: the general one and the one you want to modify. 

It should work the same way as `top_solid_infill_flow_ratio` and `bottom_solid_infill_flow_ratio`. Perhaps the term "override" is not entirely suitable here. Probably, it'd be better to replace it with another word to avoid confusion. 

But the logic of new parameters should be exactly the same as for top/bottom surface flow ratios.

valerii-bokhan commented:
@boromyr I double-checked the code and ensured that the parameters I added work the same as the existing ones.

At first, objects' and filaments' multipliers are taken into account, and then any of the others, depending on the path roles.

So the question is, how can I rename the "override" checkbox to avoid confusion?

<img width="1176" height="701" alt="image" src="https://github.com/user-attachments/assets/0e173b87-a70f-4f14-9ead-3aa1faa15056" />

boromyr commented:
Maybe it's better to follow Orca's logic and remove overrides, maybe just "enabling other flow ratios" would suffice.

boromyr commented:
Maybe only the flow for the first layer is missing; I usually increase it especially for ASA and ABS.

boromyr commented:
Actually, that value influences all the lower surfaces, not just the first layer, and it doesn't act on the brims. I found an alternative solution.
```py
{ if (layer_num==0)}
M221 S110
{endif}
```

valerii-bokhan commented:
@boromyr I double-checked the code, and I can easily add a multiplier for the first layer.

The final flow ratio for a path will be a multiplication of `object flow ratio`, `filament flow ratio`, `role-based flow ratio`, and `first layer flow ratio` if we are on the first layer.

It will follow the logic of the other multipliers, but for the first layer, infill, brim/skirt, outer, and inner walls will have different flow ratios. Is it okay, or do we want FR to be equal for all the paths on the first layer?

valerii-bokhan commented:
Okay, then I will add a multiplier for the first layer as well.

boromyr commented:
Ok, thanks, I fill it out and try it.

TomPcz commented:
Would it be possible to also adjust the flow rate for solid infill that is ironed? This would be especially useful for zero-flow ironing (PR #10350), where one might want to increase the flow. As @pi-squared-studio suggested in the discussion of that PR, one might also want to decrease the infill flow to allow for a higher ironing flow. I tried this by setting the top-surface flow ratio to 0.8 and the ironing flow to 21%, and it worked very well. However, the top-surface flow ratio applies to all top surfaces, not just the ironed ones, so modifiers must be set carefully. Having an option to adjust the flow of the ironed infill separately would resolve this issue.

valerii-bokhan commented:
@TomPcz Hey, it looks interesting, but this is a special case that would be better covered in a separate PR.

I want to look into it, but it'd also be great to merge the current PR first.

pi-squared-studio commented:
@valerii-bokhan @TomPcz 

By the way, try this PR https://github.com/SoftFever/OrcaSlicer/pull/10622
It also has a parameter for changing the flow at certain perimeters, but even this fact is not important. I noticed that if you print the perimeters after one, then the amazing accuracy of the surface remains. Perhaps this is due to the fact that when the extruded material cools, it crystallizes more evenly. There are 2 modes in this PR, this is when the outer wall is (or not) printed last. When there is an outer wall control, theoretically there are pores there, and it would be interesting to check what happens when the material a little bit overflows.

valerii-bokhan commented:
In my case, "wrinkles" look like this:
![image](https://github.com/user-attachments/assets/012b4950-b879-486f-acf0-85d9d5337ada)

These changes helped me with my things and it is not so big change, it does not touch any complex code/algorithms, etc.

valerii-bokhan commented:
Photos of recent tests (0.02mm). The only problem I have now is micro-stringing after wipe moves (I had it before having changes introduced in this PR).

![image](https://github.com/user-attachments/assets/15cf054c-0a2d-4c16-97ac-04f7f7ad888a)
![image](https://github.com/user-attachments/assets/aed747d1-01df-4cd9-8377-96344e23cd78)
![image](https://github.com/user-attachments/assets/8dab3d26-6b5b-46d7-b75f-da6e99baabc7)
![image](https://github.com/user-attachments/assets/a5e7573a-914d-40ce-a6d1-09451540f218)
![image](https://github.com/user-attachments/assets/4f17ed7d-cef6-40f6-b1e6-4386f4df88e3)

boromyr commented:
The flow of the first layer behaves strangely when you apply different flows for different objects, it is affected at the birm even you said it would not be affected.
<img width="714" height="793" alt="image" src="https://github.com/user-attachments/assets/2c50a56b-4dcb-490c-bd7a-61989ee69459" />

valerii-bokhan commented:
Yes, brims are not covered by my code. Could you please upload a project file? Or how can it be reproduced?

valerii-bokhan commented:
I see, so brims are affected by this code somehow... Thank you! I will debug it.

valerii-bokhan commented:
Fortunately, it was a quick fix, but the brim flows still look strange. I am still new to a codebase, so I don't know yet how brim's flow is being calculated.

I want to look into it separately. I hope it's ok for now.

I will make a commit with the fix soon. I need to do a couple of more tests.

<img width="1357" height="948" alt="image" src="https://github.com/user-attachments/assets/92470a69-f5dc-4987-80fd-657e3da558c0" />

<img width="1377" height="994" alt="image" src="https://github.com/user-attachments/assets/1cd872fc-6719-4606-b5f7-59a0917a9c7c" />

pi-squared-studio commented:
I repeated the test to study this phenomenon. Layer 0.02. Standard settings without increasing the flow on the walls.
As you can see, these ripples on the overhangs have remained, but they have a completely different pattern.
<img width="838" height="394" alt="image" src="https://github.com/user-attachments/assets/6655a456-84b6-4444-a7f8-a8c593eec4e1" />
<img width="671" height="918" alt="image" src="https://github.com/user-attachments/assets/8b63fcd1-6f3b-4bdf-ac79-8faeec192841" />

And now for the sad part. Initially, I didn't want to write, but since I had to study this phenomenon, I will write a report. Anyway, I do not associate the solution of this defect with an increasing flow or line width. It is related only to the materials used, printers and its calibrations, firmware and their settings (Marlin/Klipper), as well as, and most importantly, to the extruder.

I did some calculation: 
The cross-sectional area of one such layer is 0.4*0.02=0.008mm2
The cross-sectional area of the filament: 3.14*(1.75/2)^2 = 2.14mm2
The extrusion coefficient in this case is 2.14/0.008 = 300

I am using a BMG extruder with a transmission ratio of 416 steps per mm. (despite the fact that there is an internal stepper divider of 1/16).
Thus, it turns out that printing a 1mm layer takes 416/300 = 1.4 stepper steps. Or, the signal to change the extruder pitch comes approximately every 0.7mm. And since the area of the uncovered outlet section of the nozzle opens on the overhangs, this influx will be visible at the moment of the extruder step.

This leads to the conclusion that by changing the flow density or line width, we simply change the extrusion coefficient and simply blur this folded pattern. The same effect can be achieved if you make a random position of the seams.
It would also be better if you replace the extruder with a more precision one, or at least set up an increased number of steps per millimeter for the stepper.

valerii-bokhan commented:
Oh wow, thank you for your tests. And you are absolutely right, this approach "blurs" these "wrinkles/ripples". It's definitely a hack. At least, it helps to get a somewhat expected look of minis.

Currently, my value for outer walls is 1.1, and I will attempt to decrease it by adjusting other settings.

valerii-bokhan commented:
You will have to compile Orca from source if you want to use these changes.

As alternative, you can try to use "Bottom surface flow ratio" setting that is already available in Orca, but it will be applied only to infill of the first layer (it does not affect walls).

boromyr commented:
Use this alternative while you wait for this to to be united
> ```python
> { if (layer_num==0)}
> M221 S110
> {endif}
> {if (layer_num==1)}
> M221 S100
> {endif}
> ```

boromyr commented:
Do not view the increase in flow with this method because it simply inserts an M221 command that the colorful preview does not see, you can only accept that during printing the flow value is increased in the first layer and then returns normal

boromyr commented:
There is an option to hide these settings if they do not like, I don't see what annoyance can give some more control to more experienced users if all the others do not notice the difference. If only the features that all users would use half of the current settings would be added, for example I do not use all the features for multimaterials.

valerii-bokhan commented:
For clarification:

- This is a hack for my particular case (printing with 0.02mm layer height)
- In some other cases (like @boromyr mentioned), it is not a hack; it can actually be useful
- It is hidden in the UI by a checkbox - if you don't need these granular settings for FL, you will not see them until you enable the checkbox, so the UI remains clean.

xlBennylx commented:
I honestly don't get why this feature does not exist in Orca jet. I Got a brand new printer with auto z-Offset and i checked the calibration on it now several times but my first layer still has gaps between the lines and does not stick very well. 

The screenshots i posed where out of a standard profile for this printer with no changes and it clearly shows that the permitter lines from the first layer get way less flow rate. The bottom layer infill even with all flow ratios to 1 are actually as high as 1.15 if compared to the top layer. And the infill gets way to much flow ratio almost a ratio of 2. And i cant change anything about it. 
That's just crazy to me that i have to increase my first layer hight and manipulate my correct z-offset so my first layer looks OK. 

I will definitely try to compile Orca with this changes if it wont get implemented soon or switch back to cura...

valerii-bokhan commented:
I had this issue too, but I fixed it via a G-code command (in machine start block) that corrected the z offset and moved the nozzle 0.02mm closer to the plate.😅

dewi-ny-je commented:
> The screenshots i posed where out of a standard profile for this printer with no changes and it clearly shows that the permitter lines from the first layer get way less flow rate. The bottom layer infill even with all flow ratios to 1 are actually as high as 1.15 if compared to the top layer. And the infill gets way to much flow ratio almost a ratio of 2. And i cant change anything about it. That's just crazy to me that i have to increase my first layer hight and manipulate my correct z-offset so my first layer looks OK.

Looks like wither your printer profile has issues, or your printer does. The math used by Orca is sound and of you have visible issues with lines widths or gaps or extra flow, you solve that first.

Software features are not a solution for hardware or configuration issues.
Improving hardware capabilities is fine (for example vibration compensation), workarounds for hardware issues are not.

xlBennylx commented:
> > The screenshots i posed where out of a standard profile for this printer with no changes and it clearly shows that the permitter lines from the first layer get way less flow rate. The bottom layer infill even with all flow ratios to 1 are actually as high as 1.15 if compared to the top layer. And the infill gets way to much flow ratio almost a ratio of 2. And i cant change anything about it. That's just crazy to me that i have to increase my first layer hight and manipulate my correct z-offset so my first layer looks OK.
> 
> Looks like wither your printer profile has issues, or your printer does. The math used by Orca is sound and of you have visible issues with lines widths or gaps or extra flow, you solve that first.
> 
> Software features are not a solution for hardware or configuration issues. Improving hardware capabilities is fine (for example vibration compensation), workarounds for hardware issues are not.

But the print profile is exactly what orca recommended for this printer as to be the optimal settings? So when everything is calculated correctly with the flow rate and so on why is it that i get such thin lines in my first layer? And hardware cant be it just tried a print sliced with cura and a standard profile and everything is fine. So why are you hating on adding a setting that actually allows you to fine tune something orca obviously can't do? You don't have to use it if you don't like it or need it.

valerii-bokhan commented:
I completely understand your frustration.

As I understand it, the default profiles are just a starting point. They are "optimal" for abstract "perfect conditions".

You still need to run all the calibrations first. You need to dry and calibrate your filament. Print the temperature tower, perform flow ratio calibration, and calibrate the Pressure Advance, etc.

Then you need to tweak the profile settings anyway - try to make the first layer height bigger, increase the bottom surface flow ratio, etc.

Because you have your own unique conditions in the room where you use your printer, you need to adjust.

valerii-bokhan commented:
I would like to add translations for several languages in a separate commit. 

And we still need your opinion, @SoftFever

valerii-bokhan commented:
@boromyr how it's going with your tests?

BTW: In BBS they have FL for the first layer specifically and not for all bottom surfaces: https://github.com/bambulab/BambuStudio/blob/4f0f0f3e97400629205048995c77b5eedd2768e5/src/libslic3r/GCode.cpp#L5792

They cover 2 cases only, for top surfaces solid infill and FL for the first layer.

<img width="848" height="192" alt="image" src="https://github.com/user-attachments/assets/9499a725-86db-4bad-9433-97f91b30a682" />

valerii-bokhan commented:
> while doing tests to improve the seams

BTW, it's completely different topic, but maybe it can be interesting for you too (I don't know if you have the same issue with seams) - https://github.com/SoftFever/OrcaSlicer/issues/3150#issuecomment-3368478570

SoftFever reviewed (approved):
LGTM
Thank you

pi-squared-studio commented:
I slightly revised my attitude towards this PR when I development of a new calibration test.
I noticed that the affected not only by the overall flowratio, but also by the direction. So for parallel perimeter layers, the density will be the same, for a solid crossed infill the other. So calibration by line density is quite the right solution. In addition, different materials may have different filling parameters between the coils. So I think it would be more logical to install this setting in the filament settings section.

Regarding overhangs, I am also convinced that this is a problem of low-flow material with ultra-thin layers.
I also have an understanding of the process of printing overhangs, and I want to make a mod that will remove such defects as mentioned in the topic but different method. Perhaps there will be another PR on this topic soon.

AssiveMass commented:
Hi all, thanks @valerii-bokhan for implementing this.

One improvement I'd like to see regarding this is per-filament settings. IMO this option would be more useful in the filament settings instead of global print settings, because the flow is heavily material dependent.

For example, I'd like to be able to set the first layer flow ratio to 0.90 for just one filament, which has over-extrusion problems on the first layer due to much higher first layer temperature. (230 vs 200 on other layers, Polyterra PLA. Lowering first layer temps reduces adhesion drastically).

Would this be doable? Thanks!

valerii-bokhan commented:
> Hi all, thanks @valerii-bokhan for implementing this.
> 
> 
> 
> One improvement I'd like to see regarding this is per-filament settings. IMO this option would be more useful in the filament settings instead of global print settings, because the flow is heavily material dependent.
> 
> 
> 
> For example, I'd like to be able to set the first layer flow ratio to 0.90 for just one filament, which has over-extrusion problems on the first layer due to much higher first layer temperature. (230 vs 200 on other layers, Polyterra PLA. Lowering first layer temps reduces adhesion drastically).
> 
> 
> 
> Would this be doable? Thanks!

Hi! This idea was mentioned in this thread already and I completely agree with you - it's a great idea. Let me look into it. I will use ironing settings per filament as an example.

valerii-bokhan commented:
> Hi all, thanks @valerii-bokhan for implementing this.
> 
> 
> 
> One improvement I'd like to see regarding this is per-filament settings. IMO this option would be more useful in the filament settings instead of global print settings, because the flow is heavily material dependent.
> 
> 
> 
> For example, I'd like to be able to set the first layer flow ratio to 0.90 for just one filament, which has over-extrusion problems on the first layer due to much higher first layer temperature. (230 vs 200 on other layers, Polyterra PLA. Lowering first layer temps reduces adhesion drastically).
> 
> 
> 
> Would this be doable? Thanks!

Hey. It took a while, but I made a PR for the feature you requested - https://github.com/OrcaSlicer/OrcaSlicer/pull/15086
